lib/ecto/query/planner.ex

```diff
@@ -540,6 +547,17 @@ defmodule Ecto.Query.Planner do
     end
   end
 
+  @spec plan_wheres(Ecto.Query.t, atom) :: Ecto.Query.t
+  defp plan_wheres(q, adapter) do
+    wheres = q.wheres |> Enum.map(fn
+      %BooleanExpr{expr: {:in, [], [left, %Ecto.SubQuery{} = s]}} = where ->
+        %{where | expr: {:in, [], [left, plan_subquery(s, s.query.prefix || q.prefix, adapter)]}}
```

Comment:
Unfortunately this will only work if the `in` expression is at the root. So if you do this: `where: foo.bar and foo.baz in subquery(...)`, then it won't work. We would need to traverse the whole `where` to find subqueries, but this could be very expensive.

Perhaps a better option here is to add a new field to BooleanExpr called `:subqueries`. So `foo in subquery(...)` becomes at the AST level something like `foo in {:subquery, 0}` where `0` is the index of the subquery in the  `:subqueries` field.

Reply:
Addressed by c87008e and 2c7fd0fe.